doschema/jsvalidation.py

```diff
@@ -0,0 +1,320 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+import logging
+
+import jsonschema
+import six
+
+from doschema.errors import JSONSchemaCompatibilityError
+
+
+class JSONSchemaValidator(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver_factory: Resolver used to retrieve referenced schemas.
+                                 If not provided it will use default.
+        """
+        self.ignore_index = ignore_index
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def validate(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        self._validate_root(schema, ())
+
+        root = self.fields_types_dict.get(())
+
+        if not root or root.field_type != 'object':
+            err_msg = "Root field / type is not 'object' in {0}"
+            raise JSONSchemaCompatibilityError(
+                '/',
+                self.uri,
+                err_msg.format(self.uri)
+            )
+
+        for path, field in six.iteritems(self.fields_types_dict):
+            if field is None:
+                logging.warning(
+                    'No type in field ' + self._make_json_pointer(path)
+                )
+
+        self.uri = None
+        self.resolver = None
+
+    def _validate_root(self, curr_schema, curr_field):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        self._validate_collection(curr_schema, curr_field)
+
+        field = self.fields_types_dict.get(curr_field)
+
+        if "type" in curr_schema:
+            if ((curr_field in self.fields_types_dict and field is None) or
```

Comment:
The whole point of using "None" for unknown field types is to enable their definition later. More IRL.